src/test/s3/util/s3Tab.test.ts

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as fs from 'fs'
+import * as assert from 'assert'
+import * as vscode from 'vscode'
+import * as path from 'path'
+import { makeTemporaryToolkitFolder } from '../../../shared/filesystemUtilities'
+import * as testutil from '../../testUtil'
+import { S3Tab } from '../../../s3/util/s3Tab'
+import { anything, instance, mock, when, capture } from '../../utilities/mockito'
+
+describe('S3Tab', async function () {
+    const message = 'can this be read'
+    const fileName = 'test.txt'
+
+    let tempFolder: string
+    let fileUri: vscode.Uri
+    let s3Uri: vscode.Uri
+    let s3Tab: S3Tab
+    let mockedWorkspace: typeof vscode.workspace
+    let mockedWindow: typeof vscode.window
+
+    before(async function () {
+        tempFolder = await makeTemporaryToolkitFolder()
+        fileUri = vscode.Uri.file(path.join(tempFolder, fileName))
+        s3Uri = vscode.Uri.parse('s3:' + fileUri.fsPath)
+        testutil.toFile(message, fileUri.fsPath)
+
+        mockedWorkspace = mock()
+        mockedWindow = mock()
+        listTempFolder(tempFolder)
+        s3Tab = new S3Tab(fileUri, instance(mockedWindow))
+    })
+
+    it('can be opened in read-only mode', async function () {
+        when(mockedWorkspace.openTextDocument(anything())).thenReturn(Promise.resolve({ uri: s3Uri } as any))
+        when(mockedWindow.showTextDocument(anything())).thenReturn()
+
+        await s3Tab.openFileOnReadOnly(instance(mockedWorkspace))
+
+        const [uri] = capture(mockedWorkspace.openTextDocument).last()
+        assert.strictEqual((uri as vscode.Uri).fsPath, s3Uri.fsPath)
+        assert.strictEqual((uri as vscode.Uri).scheme, s3Uri.scheme)
+
+        const [showDocArgs] = capture(mockedWindow.showTextDocument).last()
+
+        assert.deepStrictEqual((showDocArgs as any).uri, s3Uri)
+    })
+
+    it('can be opened in edit mode', async function () {
+        when(mockedWorkspace.openTextDocument(anything())).thenReturn(Promise.resolve({ uri: fileUri } as any))
+        when(mockedWindow.showTextDocument(anything())).thenReturn()
+
+        await s3Tab.openFileOnEditMode(instance(mockedWorkspace))
+
+        const [uri] = capture(mockedWorkspace.openTextDocument).last()
+        assert.strictEqual((uri as vscode.Uri).fsPath, fileUri.fsPath)
+        assert.strictEqual((uri as vscode.Uri).scheme, fileUri.scheme)
+
+        const [showDocArgs] = capture(mockedWindow.showTextDocument).last()
+
+        assert.deepStrictEqual((showDocArgs as any).uri, fileUri)
+    })
+
+    it('saves changes back to s3', function () {})
```

Comment:
Nit: we don't have any tests that write to S3...I'm assuming you'll be mocking S3 in this case? If so, change the title to 'calls S3 when saving changes'.